### PR TITLE
refactor(telemetry-utils): Make `MockLogger`'s `events` property externally read-only

### PR DESCRIPTION
**Pull request by Josmithr**
There was 1 existing test case that depended on mutability to work around the fact that the mock logger clears its events after checking for a match. It accomplishes this by adding an option for whether or not to clear the events after performing a match check (default: true).

Also fixed a couple of bugs: 
* The early-out check in `matchEventStrict` would cause the events not to be cleared when they should have been.
* Errors logged by assertion messages would not include the "actual" event details, because they were cleared before the error was thrown.

### CODE REVIEW

File: packages/utils/telemetry-utils/src/mockLogger.ts [hunk 1]
```diff
@@ -62,10 +71,12 @@ export class MockLogger implements ITelemetryBaseLogger {
 	public matchEvents(
 		expectedEvents: Omit<ITelemetryBaseEvent, "category">[],
 		inlineDetailsProp: boolean = false,
+		clearEventsAfterCheck: boolean = true,
```
Comment (Josmithr) [inline packages/utils/telemetry-utils/src/mockLogger.ts:74]: TODO: document this where appropriate.

File: packages/utils/telemetry-utils/src/mockLogger.ts [hunk 2]
```diff
@@ -141,11 +156,17 @@ ${JSON.stringify(actualEvents)}`);
...
 	): boolean {
-		return (
-			expectedEvents.length === this.events.length &&
-			this.matchEvents(expectedEvents, inlineDetailsProp)
-		);
+		if (expectedEvents.length !== this._events.length) {
+			if (clearEventsAfterCheck) {
+				this.clear();
+			}
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/mockLogger.ts:164]: Nice catch. Worth a new unit test :).

**Review by alexvy86 — COMMENTED** [review]
LGTM, but I'd like us to get some clarity on how/when we might need to communicate a breaking change to alpha APIs to our partners. I think we still want to provide some guarantees there but I'm not clear on what exactly they are or how breaking changes are communicated when they need to happen.

File: packages/utils/telemetry-utils/src/mockLogger.ts (hunk 2 above)
Comment (Josmithr) [reply packages/utils/telemetry-utils/src/mockLogger.ts:168]: Done 🙂

**Comment by Josmithr** [comment]
> LGTM, but I'd like us to get some clarity on how/when we might need to communicate a breaking change to alpha APIs to our partners. I think we still want to provide some guarantees there but I'm not clear on what exactly they are or how breaking changes are communicated when they need to happen.

Until our release candidate branch is cut (tomorrow), we're still just depending on changesets. In this case, I think the change is safe to make. I've added a changeset doc explaining what users should do if some behavior breaks, but I don't think it will be an issue.

File: packages/utils/telemetry-utils/src/mockLogger.ts [hunk 3]
```diff
@@ -26,8 +26,14 @@ import type {
  * @alpha
  */
 export class MockLogger implements ITelemetryBaseLogger {
-	// TODO: don't expose mutability to external consumers
-	public events: ITelemetryBaseEvent[] = [];
+	/**
+	 * Gets an immutable copy of the events logged thus far.
```
Comment (Josmithr) [inline packages/utils/telemetry-utils/src/mockLogger.ts:30]:
```suggestion
	 * Gets an immutable copy of the events logged thus far.
	 *
	 * @remarks Note: these contents are cleared after calling any of the match checking / assertion methods on this class.
```

File: packages/utils/telemetry-utils/src/mockLogger.ts [hunk 4]
```diff
@@ -62,30 +71,45 @@ export class MockLogger implements ITelemetryBaseLogger {
...
 		);
 		// How many expected events were left over? Hopefully none.
 		const unmatchedExpectedEventCount = expectedEvents.length - matchedExpectedEventCount;
 		return unmatchedExpectedEventCount === 0;
 	}
 
 	/**
-	 * Asserts that matchEvents is true, and prints the actual/expected output if not.
+	 * Asserts {@link MockLogger.matchEvents} is `true` for the given events.
+	 * @param disallowedEvents - The events expected to not appear.
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/mockLogger.ts:94]:
```suggestion
	 * @param expectedEvents - The events expected to appear.
```

File: packages/utils/telemetry-utils/src/mockLogger.ts [hunk 5]
```diff
@@ -107,24 +131,37 @@ ${JSON.stringify(actualEvents)}`);
...
 			inlineDetailsProp,
+			clearEventsAfterCheck,
 		);
 		return matchedExpectedEventCount > 0;
 	}
 
 	/**
-	 * Asserts that matchAnyEvent is true, and prints the actual/expected output if not.
+	 * Asserts {@link MockLogger.matchAnyEvent} is `true` for the given events.
+	 * @param disallowedEvents - The events expected to not appear.
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/mockLogger.ts:146]:
```suggestion
	 * @param expectedEvents - The events expected to appear.
```

File: packages/utils/telemetry-utils/src/mockLogger.ts [hunk 6]
```diff
@@ -141,27 +178,46 @@ ${JSON.stringify(actualEvents)}`);
...
+		}
+
+		// `events` will be cleared by the below check if requested.
+		return this.matchEvents(expectedEvents, inlineDetailsProp, clearEventsAfterCheck);
 	}
 
 	/**
-	 * Asserts that matchEvents is true, and prints the actual/expected output if not
+	 * Asserts {@link MockLogger.matchEvents} is `true` for the given events.
+	 * @param disallowedEvents - The events expected to not appear.
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/mockLogger.ts:202]:
```suggestion
	 * @param expectedEvents - The events expected to appear.
```

**Review by alexvy86 — APPROVED** [review]
LGTM, just a couple of doc corrections. I think one of the functions that was already documented has more detail about the events when we do expect them to appear, but I didn't check if all of them go through the same code path and the same behavior applies (e.g. if the events are expected in strict order in every case).

File: packages/utils/telemetry-utils/src/mockLogger.ts (hunk 4 above)
Comment (Josmithr) [reply packages/utils/telemetry-utils/src/mockLogger.ts]: Nice catches! Thanks!!

**Pull request merged**